### ISSUE TITLE
Header and footer embedded in the html page

### ISSUE THREAD

**Issue by tbruyelle** [wkhtmltopdf/wkhtmltopdf#3053]
Instead of passing header and footer in the command line parameters, I wonder if it's possible to have them embedded in the html source page.

The benefit would be to have a single html page with one header and one footer, whereas the pdf would have header and footer on each page.

**Comment by sebalis** [comment]
This is a feature I would urgently need in order to replace dompdf as the PDF generator library within CiviCRM, a Free Software aimed at NGOs for their customer (or “constituent”) relationship management. In CiviCRM you can define any number of message templates which are then filled with data from contacts or other entities. The templates are stored in the software’s database and as they are filled out with individual data they produce temporary files. So it is impossible to create separate files for headers or footers and pass them as command line options because CiviCRM only creates one file (and it’s temporary so we would not know with file name to use with ```--header-html```). Also, CiviCRM incorporates wkhtmltopdf only through a single configuration setting containing the command line so there is no way of accomodating different header/footer files per individual template, which is of course what would be needed.

In dompdf, header and footer content is marked as such simply by using the CSS declaration ```position: fixed``` together with other declarations (containing the necessary measures for placing the elements on the pages). The HTML elements &lt;header&gt; and &lt;footer&gt; could be given these CSS settings for a very natural way of generating headers and footers from within the main document. (I looked for documentation and found [this Github issue](https://github.com/dompdf/dompdf/issues/1190).)

Any chance of seeing this in wkhtmltopdf? Thanks for your consideration.

**Comment by sebalis** [comment]
In case you’re wondering why I want to move away from dompdf, I won’t go into unnecessary details but it has several issues. Using wkhtmltopdf seems to be a great way forward, it fixes a few bugs and is much better in performance. One of those issues is that while the mechanism I described above does exist in dompdf it doesn’t actually work in CiviCRM, perhaps because they are using an old version (but I don’t know).

**Comment by PhilterPaper** [comment]
If I understand your description, CiviCRM is creating a dompdf-oriented HTML source that includes `<header>` and `<footer>` with CSS positioning for page headers and footers. Presumably this is fixed content (no page numbers, etc.). I take it that this setup doesn't work with wkHTMLtoPDF (possibly too advanced a level of HTML and CSS for WebKit to handle?). Is this correct? And you **don't** have access to the temporary HTML file's name? About the only thing I could think of, if you _can_ set the command line, would be to wrap wkHTMLtoPDF in a wrapper script that accepts the temporary HTML file, looks inside to extract the header and footer information (possibly removing it from the HTML), writing some temporary HTML files to use with --header-html and --footer-html, and finally invoking wkHTMLtoPDF on the command line to run it with all these temporary files. 

If dompdf doesn't support page numbers, etc., in its headers and footers, the extraction process could reformat their HTML to use page numbers, etc. Just be careful when handling temporary files to not give them fixed names which would cause problems when multiple users' copies collide.

It's interesting that you state that you want to drop dompdf in favor of wkHTMLtoPDF because the latter is less buggy and better performing. I've seen a number of posts stating just the opposite -- that they are leaving wkHTMLtoPDF in favor of dompdf because it's less buggy and faster. I guess the grass is always greener on the other side!

I hope this has given you some ideas for avenues to explore.

**Comment by sebalis** [comment]
Yes, that is roughly right. Although the HTML that CIviCRM creates is not specific for dompdf, it’s almost entirely under user control (to be precise, the content of the body tag is), so the HTML can be geared towards whatever is in use.

Page numbers and total number of pages would be desirable too of course, and I do wonder if this can be achieved with wkhtmltopdf. I have just checked and found some rather complicated looking JavaScript constructions – I hope that they work. In dompdf there is a more elegant solution for the page number: include a span with, say, ```class="pageNumber"``` in the HTML and add the page number using CSS:  ```.pageNumber:after { content: counter(page); }``` But there is no such solution for the total number of pages.

Yes, we do not know the temporary file’s name, not even through some placeholder in the config setting that tells CIviCRM which command to run – and yes, I did consider a wrapper script much in the way you describe, @PhilterPaper. But it would be hard to make that generic enough, as it would have to extract both the header section and any style tags contained in the body. Not impossible but cumbersome. And it doesn’t really make sense – conceptually the header and footer is part of the same document, so that is where it should be defined. (How does positioning work anyway? Much more logical to place a HTML element and make room for it using the single document’s CSS, rather than leaving white space in one document and filling it from another one.) What I like about the way this _would_ work in dompdf (if it did work as documented) is that again it is declarative and driven by the HTML’s content, without having to revert to externals such as other documents, the details of the command line, environment variables, or even PHP (which would be the way to include page totals in dompdf, this would not only be very cumbersome but also not update-proof).

On the other side, wkhtmltopdf definitely has performance benefits, there is clear evidence for that both in our own experience and in blog articles. So it would be perfect if some of the good things dompdf has to offer, namely the declarative way of defining headers/footers and including page numbers/totals, could find its way into wkhtmltpdf.

**Comment by sebalis** [comment]
Any chance of adopting a declarative approach so that headers/footers and page numbers could be defined by the main document’s HTML/CSS content rather than by command lines or other externals, @ashkulz? :-)

**Comment by PhilterPaper** [comment]
Yes, wkHTMLtoPDF's header and footers include page and total page substitutions:
> As an example specifying --header-right "Page [page] of [toPage]", will result in the text "Page x of y" where x is the number of the current page and y is the number of the last page, to appear in the upper left corner in the document.

Someone should try running wkHTMLtoPDF with `<header>` and `<footer>`, and use CSS to position absolutely, and see what it does. If that's sufficient, you could use the existing command-line header and footer to add "Page x of y" on top of that.

Perhaps if you can't get an HTML file directly out of CiviCRM, you could use a wrapper script to extract the HTML file for further processing (before feeding to wkHTLtoPDF). If the HTML header/footer produced for dompdf works OK with wkHTMLtoPDF, you're largely done (except for Page x of y, using command line header or footer. If you have to extract the `<header>` or `<footer>` text into files because wkHTMLtoPDF can't handle it, that's a nuisance but not an insurmountable problem. Note that wkHTMLtoPDF has a large number of substitution fields:
```
 * [page]       Replaced by the number of the pages currently being printed
 * [frompage]   Replaced by the number of the first page to be printed
 * [topage]     Replaced by the number of the last page to be printed
 * [webpage]    Replaced by the URL of the page being printed
 * [section]    Replaced by the name of the current section
 * [subsection] Replaced by the name of the current subsection
 * [date]       Replaced by the current date in system local format
 * [isodate]    Replaced by the current date in ISO 8601 extended format
 * [time]       Replaced by the current time in system local format
 * [title]      Replaced by the title of the of the current page object
 * [doctitle]   Replaced by the title of the output document
 * [sitepage]   Replaced by the number of the page in the current site being converted
 * [sitepages]  Replaced by the number of pages in the current site being converted
 ```
so you may not even _need_ any of the dompdf header/footer capabilities -- it might all be done in wkHTMLtoPDF. Can you shut off headers and footers in CiviCRM?

**Comment by sebalis** [comment]
As I said, I can pretty much produce whatever HTML I want from CiviCRM. So it’s not HTML _for dompdf_ but whatever HTML I need for the tools I use. But it is definitely one single HTML file. Also I can not use an extra command line option to add more content as this would be hard to maintain and, more importantly, impossible to individualise. We are talking about a system that can have any number of templates all with their own design needs – it would be absolutely impossible to cover all that with an option that inserts a fixed text.

Extracting an HTML is not the issue. I would need the HTML that is produced after variable subsitution – we’re talking mail merge functionality here – and that is what the wrapper script would be given. Then all the ideas you kindfully suggested could be tried. But that’s not my point. I am hoping to persuade the wkhtmltopdf developers to adopt or at least consider a different approach – where one HTML file has all the information required to produce headers and footers – an approach for which I used the word declarative earlier.

**Comment by PhilterPaper** [comment]
See also #5015 in case that gives you any ideas.
